### PR TITLE
Bugfix/refact user add+update forms + fix user details forms 

### LINKED ISSUE #186: Users screen - Validation on username don't work
needs #243

### LINKED ISSUE #188: Users page - Validation email field doesn't match the requirements
needs #243

### LINKED ISSUE #243: Refact user forms - starting with "Can't update users details"
closes issue #189

### PR DESCRIPTION
**Pull request by tulitheprogrammer**
The forms now inherit from the same base class (DRY) 
+ bug fixes: 

1. Saved user role + class show on drop downs title
2. role is being saved on update form
3. validations are the same as most behaviors on both form types .
this PR closes #243 and closes #186 closes #188

### CODE REVIEW

File: src/app/pages/user/dialogs/base/user-details-form/user-details-form.html [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+          <mat-error *ngIf="EmailFormControl.hasError('required')">אימייל הינו שדה חובה</mat-error>
+          <mat-error *ngIf="EmailFormControl.hasError('email')">אימייל לא חוקי</mat-error>
+        </mat-form-field>
+      </div>
+      <div class="msw-form-row">
+        <mat-form-field color="primary">
+          <mat-select placeholder="כיתה" formControlName="class" [(ngModel)]="data.class" name="class" [compareWith]="compareClass">
+           <mat-option disabled>בחר</mat-option>
+           <mat-option *ngFor="let class1 of classes" [value]="class1">
+              {{class1.name}}
```
Comment (dimkinv) [inline src/app/pages/user/dialogs/base/user-details-form/user-details-form.html]: hey, `class1`?

File: src/app/pages/user/dialogs/base/user-details-form/user-details-form.scss [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  align-items: center;
+  text-align: center;
+}
+
+.mat-dialog-actions button {
+  text-align: center;
+}
+
+.mat-form-field {
+  //  background-color: aqua;
```
Comment (dimkinv) [inline src/app/pages/user/dialogs/base/user-details-form/user-details-form.scss]: throughout this file, can you please remove commented code?

File: src/app/pages/user/dialogs/base/user-details-form/user-details-form.spec.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+      role: UserType.PRINCIPLE,
+      username: 'a',
+      _id: 0,
+      class: new Class(),
+    };
+
+    TestBed.configureTestingModule({
+      imports: [
+        ReactiveFormsModule,
+        // MatFormFieldModule,
```
Comment (dimkinv) [inline src/app/pages/user/dialogs/base/user-details-form/user-details-form.spec.ts]: same here, can you please remove commented code?

File: src/app/pages/user/dialogs/base/user-details-form/user-details-form.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  close() {
+    this.cancel.emit();
+  }
+
+  submitForm() {
+    this.formSubmit.emit(this.form.value);
+  }
+
+  compareClass(c1: Class, c2: Class): boolean {
+    return c1 && c2 ? c1._id === c2._id : c1 === c2;
```
Comment (dimkinv) [inline src/app/pages/user/dialogs/base/user-details-form/user-details-form.ts]: why you explain this function?

Comment (tulitheprogrammer) [reply src/app/pages/user/dialogs/base/user-details-form/user-details-form.ts]: Hi  @dimkinv,
This function is used for showing the right initial value on class select box

File: src/app/pages/user/dialogs/base/user-details-form/user-details-form.scss (hunk 2 above)
Comment (tulitheprogrammer) [reply src/app/pages/user/dialogs/base/user-details-form/user-details-form.scss]: Done

File: src/app/pages/user/dialogs/base/user-details-form/user-details-form.spec.ts (hunk 3 above)
Comment (tulitheprogrammer) [reply src/app/pages/user/dialogs/base/user-details-form/user-details-form.spec.ts]: Done

File: src/app/pages/user/dialogs/base/user-details-form/user-details-form.html (hunk 1 above)
Comment (dimkinv) [reply src/app/pages/user/dialogs/base/user-details-form/user-details-form.html]: I agree about distinguishing. But not with class1. Let's find better naming for that

File: src/app/pages/user/dialogs/base/user-details-form/user-details-form.ts (hunk 4 above)
Comment (dimkinv) [reply src/app/pages/user/dialogs/base/user-details-form/user-details-form.ts]: Yes. What I'm trying to understand is why so many assertions needed? Does this mean that the data is not consistent?

**Pull request merged**